### ISSUE TITLE
[Tester Wanted] Test AbetterRoutePlanner connexion

### ISSUE THREAD

**Issue by flobz** [flobz/psa_car_controller#73]
Hello,
I implement abetterrouteplanner connexion in [feature-abrp](https://github.com/flobz/psa_car_controller/tree/feature-abrp) branch.
It send car data to abetterrouteplanner to avoid specify manually these information when you use abrp in realtime.
I documented the procedure [here](https://github.com/flobz/psa_car_controller/blob/feature-abrp/docs/abrp.md).
@jlayec @farom57 @Mrgove10 @xammmue @Flodu31 or anyone else can test it ?
I don't have access to my car right now...
Thanks

**Comment by Flodu31** [comment]
Hello,

no problem to try it.
I’ll have same functions as I have with the current one?
Or do I need to deploy it in another Pi?

thanks.
Florent

**Comment by flobz** [comment]
Hello,
It's the same with more functionality.
just do a `git checkout feature-abrp `
Thank for your help.

**Comment by CalicodesIles** [comment]
Thank you flobz.

Do you know if more information is provided with this feature-abrp branch to ABRP compared to Tronity.io (which is not free) ?

**Comment by flobz** [comment]
I'didnt test trionity, but same level of informations should be be given because the source is the same : psa api.

**Comment by flobz** [comment]
The code :
{"utc": int(datetime.timestamp(energy.updated_at)),
                       "soc": energy.level,
                       "speed": car.status.kinetic.speed,
                       "car_model": car.get_abrp_name(),
                       "current": car.status.battery.current,
                       "is_charging": energy.charging.status == "InProgress",
                       "lat": car.status.last_position.geometry.coordinates[1],
                       "lon": car.status.last_position.geometry.coordinates[0],
                       "power": energy.consumption
                       }

**Comment by Flodu31** [comment]
Ok, I've ABRP with value none:

```
Your vehicles: ["{'vin': 'vin', 'vehicle_id': 'id', 'label': None, 'brand': 'Peugeot', 'battery_power': 46, 'fuel_capacity': 0, 'max_elec_consumption': 70, 'max_fuel_consumption': 0, 'abrp_name': None}"]
```

I can't find the generic part  on ABRP, like in your screenshot:

https://github.com/flobz/psa_car_controller/raw/feature-abrp/docs/abrp.png

Florent

**Comment by Flodu31** [comment]
Hello @flobz 

I did everything but I can't see the control tab in my interface:

![image](https://user-images.githubusercontent.com/15648175/114148893-93096d00-991a-11eb-824e-615de529fc0d.png)

Any idea?

Thanks.
Florent

**Comment by flobz** [comment]
Are you sure that you use the version on the feature-abrp branch ?
Do you successfully add the abrp token ? It should be in the config file if it works.

**Comment by Flodu31** [comment]
Yes I'm sure, I did the following command:

```
git clone -b feature-abrp https://github.com/flobz/psa_car_controller.git ./
```

I added the ABRP token with the following command:

```
http://10.0.253.5:5000/abrp?token=my-token-?-?-?
```

Finally, in the config file I have this:

```
2021-04-09 11:02:00,690 :: INFO :: ["{'vin': 'vin', 'vehicle_id': 'id', 'label': None, 'brand': 'Peugeot', 'battery_power': 46, 'fuel_capacity': 0, 'max_elec_consumption': 70, 'max_fuel_consumption': 0, 'abrp_name': None}"]
```

**Comment by flobz** [comment]
what is the result of http://10.0.253.5:5000/abrp?token=my-token-?-?-? ?

**Comment by Flodu31** [comment]
Ok I stopped the server, restart it, executed the url with abrp and I had the correct webpage shown. I then go back to the original  URL and I have now the control tab. I activated it.

What do I ned to put in the null value?

{"Peugeot;e-2008 SUV (alpha)": "peugeot:e2008:20:48"}

Thanks.

**Comment by flobz** [comment]
"peugeot:e2008:20:48"

**Comment by Flodu31** [comment]
Ok done.
I have this now:

```
INFO :: ["{'vin': 'vin', 'vehicle_id': 'id', 'label': 'peugeot:e2008:20:48', 'brand': 'Peugeot', 'battery_power': 46, 'fuel_capacity': 0, 'max_elec_consumption': 70, 'max_fuel_consumption': 0, 'abrp_name': None}"]
```

I have this in logs too:

```
2021-04-09 11:12:03,439 :: ERROR :: Traceback (most recent call last):
  File "/usr/local/src/psa_car_controller/web/abrp.py", line 28, in call
    "speed": car.status.kinetic.speed,
AttributeError: 'NoneType' object has no attribute 'speed'
```

If it can help.

**Comment by flobz** [comment]
It should be fixed now. Thanks for you feedback.

**Comment by Flodu31** [comment]
Do I need to redeploy all the app?

**Comment by flobz** [comment]
No just do a git pull

**Comment by flobz** [comment]
* and restart

**Comment by Flodu31** [comment]
Good, but I have now this error:

```
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]: 2021-04-09 12:47:03,032 :: ERROR :: Traceback (most recent call last):
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]:   File "/usr/local/src/psa_car_controller/web/abrp.py", line 31, in call
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]:     "car_model": car.get_abrp_name(),
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]:   File "/usr/local/src/psa_car_controller/Car.py", line 109, in get_abrp_name
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]:     raise ValueError("ABRP model is not set")
Apr 09 12:47:03 FLOAPP-PEUGEOT psa_car_controller[18021]: ValueError: ABRP model is not set
```

In my cars.json, I have "abrp_name": null
Florent

**Comment by flobz** [comment]
1.Stop the server
2. Check abrp_name in cars.json, it's probably nulll

**Comment by Flodu31** [comment]
Yesn abrp_name is null

**Comment by flobz** [comment]
It should be: "abrp_name":  "peugeot:e2008:20:48"
Also label should be "e-208"

**Comment by Flodu31** [comment]
Ok I didn't understand, but now it's clear :)
No more error apparently.

**Comment by flobz** [comment]
Ok you just have to drive to test it then ! :)

**Comment by Flodu31** [comment]
Edit, it doesn't work :( 
I started the server, and I have the same error message as before.
And the arbp_name value is set  to null again  :(

**Comment by flobz** [comment]
ok should be fixed now , could you :
git pull
stop the server
set again abrp_name in cars.json
restart

**Comment by Flodu31** [comment]
Ok better.
I have now this error:
```
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]: 2021-04-09 14:53:03,418 :: ERROR :: kinetic not available from api
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]: 2021-04-09 14:53:03,419 :: ERROR :: last_position not available from api
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]: 2021-04-09 14:53:03,424 :: ERROR :: Traceback (most recent call last):
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]:   File "/usr/local/src/psa_car_controller/web/abrp.py", line 34, in call
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]:     "lat": car.status.last_position.geometry.coordinates[1],
Apr 09 14:53:03 FLOAPP-PEUGEOT psa_car_controller[18860]: AttributeError: 'NoneType' object has no attribute 'geometry'
```

**Comment by flobz** [comment]
And now ?

**Comment by flobz** [comment]
Ok and now ?

**Comment by Flodu31** [comment]
At this time, it is better. I don't have any error message, thanks :)

**Comment by Flodu31** [comment]
Hey @flobz 
I have this:
![image](https://user-images.githubusercontent.com/15648175/114261487-1e4a3780-99db-11eb-8d06-5c8d0eb085a5.png)
What can I check more? Because to see travels, apparently, I need to have a premium account.
Florent

**Comment by flobz** [comment]
The api status should be "connected"...
It doesn't seemto work since yesterday at 16h58

**Comment by flobz** [comment]
Maybe the localization is mendatory

**Comment by xammmue** [comment]
Hi, just tried it out now because I had to reset my pin.

After setting the correct label in car.json (2 times because I forgot to pull) it did not work out of the box but that could be on me. After I sent a command to the car (which was inactive since April 5) and a restart it worked the link to ABRP seems to work. API status is connected and last updated shows 10.04.2021 12:02:16.

**Comment by flobz** [comment]
Ok cool ! Please let me know if it work well during a trip.

**Comment by flobz** [comment]
@xammmue and @Flodu31 can you give me the beginning of your vin ?
According to this page [https://en.wikibooks.org/wiki/Vehicle_Identification_Numbers_(VIN_codes)/Peugeot/VIN_Codes](https://en.wikibooks.org/wiki/Vehicle_Identification_Numbers_(VIN_codes)/Peugeot/VIN_Codes) we can determine model 
with the first 10 char ! 
Like this we no longer have to set it manually.
Thanks

**Comment by Flodu31** [comment]
For me, the 10 first are:

![Uploading 38CB7738-CA75-4573-9FDD-6BA4E40BD504.jpeg…]()

**Comment by xammmue** [comment]
VXKUHZKXZL

**Comment by CalicodesIles** [comment]
@flobz Would you need mine (e208 from 01/2020) or yours would be sufficient) ?

**Comment by flobz** [comment]
> @flobz Would you need mine (e208 from 01/2020) or yours would be sufficient) ?
Please give me yours also.

**Comment by CalicodesIles** [comment]
Mine begins with VR3UHZKXZL

**Comment by xammmue** [comment]
I just saw the otermop API complains about the sent request:
` {"status": "ok", "missing": "Missing telemetry: power, speed, "}`
Adding `or 0` for speed and power (lines 39 and 36 in abrp.py) should solve the issue

**Comment by xammmue** [comment]
I also noticed that ABRP now also displayed not connected for me. It also still displayed 12:09 als last refresh time - same as in the myOpel App.
I then refreshed the state once by sending a command to the car from the phone. After that the last updated in myOpel changed to 16:17 and ABRP also received that 16:17 with the next request and displayed connected.

So not connected does not entirely mean that it does not work, just that the data from the car is old

**Comment by flobz** [comment]
> So not connected does not entirely mean that it does not work, just that the data from the car is old

Thanks I didn't know that.
> I just saw the iternio API complains about the sent request:
But it works ?

**Comment by xammmue** [comment]
I think it still works, I originally thought that could be the reason for the not connected state, but it didn't change after I made the program send 0 instead of None.

**Comment by Flodu31** [comment]
I can confirm @xammmue comment. We used the car this afternoon, and the update time changed:

![Uploading 01844B77-254A-451A-9884-153826F4838F.png…]()

**Comment by CalicodesIles** [comment]
Your latest commit seems to work correctly with model identification based on VIN.
Connection with ABRP is successful, data is correctly collected, unfortunately I can't go further with my Télémaintenance option issues.

**Comment by CalicodesIles** [comment]
> Ok done.
> I have this now:
> 
> ```
> INFO :: server start
> WARNING :: Can't get car model please check cars.json
> INFO :: ["{'vin': 'vin', 'vehicle_id': 'id', 'label': 'peugeot:e2008:20:48', 'brand': 'Peugeot', 'battery_power': 46, 'fuel_capacity': 0, 'max_elec_consumption': 70, 'max_fuel_consumption': 0, 'abrp_name': None}"]
> ```
> 
> I have this in logs too:
> 
> ```
> 2021-04-09 11:12:03,439 :: ERROR :: Traceback (most recent call last):
>   File "/usr/local/src/psa_car_controller/web/abrp.py", line 28, in call
>     "speed": car.status.kinetic.speed,
> AttributeError: 'NoneType' object has no attribute 'speed'
> ```
> 
> If it can help.

Just for people who tried to upgrade their install. I had some trouble based on this comment, and there is a mistake here : 
"label" value should be your car model (for example "e-2008") and abrp_name should be set to "peugeot:e2008:20:48" instead of None

**Comment by flobz** [comment]
>     * "abrp_name" should be set to "peugeot:e2008:20:48" instead of None

If you have a 208 it's not this one...
thank you all for your test!

**Comment by xammmue** [comment]
> So not connected does not entirely mean that it does not work, just that the data from the car is old
I tried it out with a timer and it seems like the API state will automatically change to Not connected when the last upadte is older than 30 minutes.

Possible solution would be to sent the current time instead of the time of the last update by the car.
Decision is just having the actual last update time vs having the Api status display connected.

Sidenote: Tronity also shows a last updated date on their page (and probably also in abrp but I did not test this out), but they seem to always display the time of the request.

**Comment by xammmue** [comment]
I tried it during a trip yesterday (only half of it as those communication services activated mid trip and no more data got sent via API).
I used it together with the ABRP Beta Android Auto feature so that I was able to display my position and SoC on the car display. SoC didn't change (that part of the trip was downhil for a good part and also the myOpel SoC was at 70% upon arrival).
The car position did only update 2 or 3 times during that < 10 minute trip. I think this is because of the interval the data is queried from the PSA servers. This makes the feature virtually useless for navigation purposes but ABRP has the use device location feature which then uses the phones (or whatever gadget you run abrp on) GPS to fetch the current position. More frequent position updates (when moving) could resolve this problem but could also be a risky behavior as we don't know anything about rate limits on the PSA API as far as I am aware.

After I was back home it seemed like I could establish a connection to the car again via the MyOpel APP and when I opened the ABRP app again, the correct SoC was displayed. So this seems to work. Also the ABRP connection status switched to connected a few minutes after I started my car which seems to confirm the previous observations.
